### PR TITLE
bears/haskell: Add GhcModBear

### LINKED ISSUE #297: Add Haskell ghc-mod Bear
Add [ghc-mod](https://hackage.haskell.org/package/ghc-mod) support for `coala-bears`

### PR DESCRIPTION
**Pull request by vijeth-aradhya**
Add GhcModBear for syntax checking
for Haskell files (.hs)

Closes #297

### CODE REVIEW

File: bears/haskell/GhcModBear.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@linter(executable='ghc-mod',
+        output_format='regex',
+        output_regex=r'(?P<filename>.+):(?P<line>\d+):(?P<column>\d+):'
+                     r'(?P<message>.+)')
+class GhcModBear:
+    """
+    Syntax checking with ghc for haskell files (.hs)
+
+    Check out https://hackage.haskell.org/package/ghc-mod for more information!
```
Comment (Nosferatul) [inline bears/haskell/GhcModBear.py]: the link should be between <>

File: bears/haskell/GhcModBear.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    DistributionRequirement)
+
+
+@linter(executable='ghc-mod',
+        output_format='regex',
+        output_regex=r'(?P<filename>.+):(?P<line>\d+):(?P<column>\d+):'
+                     r'(?P<message>.+)')
+class GhcModBear:
+    """
+    Syntax checking with ghc for haskell files (.hs)
```
Comment (Nosferatul) [inline bears/haskell/GhcModBear.py]: I'm not sure if we need `(.hs)` here

File: bears/haskell/GhcModBear.py (hunk 1 above)
Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Alright! :smile:

File: bears/haskell/GhcModBear.py (hunk 2 above)
Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Hm I need to specify the file extension
Any ideas on how? []?

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Any reason why CircleCI is failing? :/

File: bears/haskell/GhcModBear.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+@linter(executable='ghc-mod',
+        output_format='regex',
+        output_regex=r'(?P<filename>.+):(?P<line>\d+):(?P<column>\d+):'
+                     r'(?P<message>.+)')
+class GhcModBear:
+    """
+    Syntax checking with ghc for haskell files.
+
+    See <https://hackage.haskell.org/package/ghc-mod for more information>
+    for more information!
```
Comment (sims1253) [inline bears/haskell/GhcModBear.py]: double "for more information"

**Comment by sims1253** [comment]
your builds are failing :(

File: tests/haskell/GhcModBearTest.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+bad_file3 = """
+lastButOne :: a -> a
+lastButOne xs | (length xs >= 2) = xs !! ((length xs) - 2)
+              | otherwise = error "Not possible"
+
+main :: IO()
+main = do
+    print(lastButOne(["asd", "ASd"]))
+"""
+
```
Comment (Mixih) [inline tests/haskell/GhcModBearTest.py]: hmm let's use a test class with a skipDecorator so we know whether it's skipping because of a depend or some internal logic error. See http://api.coala.io/en/latest/Developers/Testing_Bears.html

Comment (Mixih) [inline tests/haskell/GhcModBearTest.py]: actually that is unnecessary... Just install it properly in ci

**Comment by Mixih** [comment]
Circle dep hell.... Check to see if something is different about the apt stuff on circle

**Comment by vijeth-aradhya** [comment]
I have looked up a good version `ghc-mod-3.1.4-1` apt package for circle, and travis also has a similar version now :+1: I tried installing the latest version `ghc-mod-5.0.6` through apt itself ( not cabal ), but it was failing probably because we need to pump up other dependencies as well :/
But, yeah I have simplified it as much as possible :) should be fine now @sils

**Comment by sils** [comment]
oh this looks much more beautiful!

File: bears/haskell/GhcModBear.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    See <https://hackage.haskell.org/package/ghc-mod> for more information!
+    """
+
+    LANGUAGES = {'Haskell'}
+    REQUIREMENTS = {DistributionRequirement(apt_get='ghc-mod')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/dzdlgzwr7ailo5vilrojnwb11'
```
Comment (sils) [inline bears/haskell/GhcModBear.py]: this shows a warning, that's not nice. I think @Makman2 introduced this and it's trying to tell you to replace `(?P<filename>.+):` with `.+:`.

File: tests/haskell/GhcModBearTest.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from bears.haskell.GhcModBear import GhcModBear
+from coalib.testing.LocalBearTestHelper import verify_local_bear
+
+
+good_file1 = """
```
Comment (sils) [inline tests/haskell/GhcModBearTest.py]: every testfile should have a purpose, give them proper names so people see the purpose in the name and they don't have to thing so much ;)

File: bears/haskell/GhcModBear.py (hunk 5 above)
Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Done :+1: Wait isn't (?P<filename>.+) there :open_mouth:

File: tests/haskell/GhcModBearTest.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+"""
+
+# Takes a file and prints the number of lines
+good_file2 = """
+main :: IO()
+main = interact lineC
+    where lineC input = show (length(lines input))
+"""
+
+# Same program but paranthesis is missing
```
Comment (sils) [inline tests/haskell/GhcModBearTest.py]: paranthesis -> parenthesis

File: tests/haskell/GhcModBearTest.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+bad_file2 = """
```
Comment (sils) [inline tests/haskell/GhcModBearTest.py]: can we still give them proper names? Enumerating isn't helpful. Good names are better than comments because they appear wherever the variable is used

File: bears/haskell/GhcModBear.py (hunk 5 above)
Comment (sils) [inline bears/haskell/GhcModBear.py]: warning gone? You gotta update the asciinema.

File: tests/haskell/GhcModBearTest.py (hunk 4 above)
Comment (aptrishu) [inline tests/haskell/GhcModBearTest.py]: No really necessary IMO.

**Comment by vijeth-aradhya** [comment]
@sils Are the testcase name okay now :no_mouth:

File: .ci/deps.sh [hunk 9]
```diff
@@ -27,7 +27,6 @@ case $CIRCLE_BUILD_IMAGE in
     # Work around lack of systemd on trusty, which xenial's lxc-common expects
     echo '#!/bin/sh' | sudo tee /usr/bin/systemd-detect-virt > /dev/null
     sudo chmod a+x /usr/bin/systemd-detect-virt
-
```
Comment (adtac) [inline .ci/deps.sh]: Unrelated change

**Comment by sils** [comment]
ack https://github.com/coala/coala-bears/pull/1223/commits/f976f0091d39324b899f5d84f06d6f14419e0559

**Comment by sils** [comment]
Rebase needed, then we can merge it :)

File: tests/haskell/GhcModBearTest.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+main = do
+    print(lastButOne(["asd", "ASd"]))
+"""
+
+GhcModBearTest = verify_local_bear(
+    GhcModBear,
+    valid_files=(hello_world_file, last_but_one_list, count_lines),
+    invalid_files=(not_string_3, count_lines_bracket_missing,
+                   last_but_one_list_type_error),
+    tempfile_kwargs={'suffix': '.hs'})
```
Comment (sils) [inline tests/haskell/GhcModBearTest.py]: if you want to get bonus points, write a proper testcase, that checks for the right result being yielded and not any

Comment (vijeth-aradhya) [inline tests/haskell/GhcModBearTest.py]: @sils Yeah that's a good idea!!
Can we do that via another issue, since it's just adding the testcase? I'll take it up myself, if no one does in probably a day or so :)

**Comment by sils** [comment]
in this case let's do it right away, we're trying to improve our testsuite
and today's a good day to start enforcing better tests ;)

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala.io/ - http://viperdev.io/ - http://gitmate.com/

File: .ci/deps.sh [hunk 11]
```diff
@@ -94,6 +94,15 @@ if [[ -z "$(which hlint)" ]]; then
   sudo dpkg -i $hlint_deb
 fi
 
+# Get ghc-mod if not installed
+if [[ -z "$(which ghc-mod)" ]]; then
```
Comment (Mixih) [inline .ci/deps.sh]: @vijeth-aradhya can we elimintate this if we upgrade circle to a newer version? if so, please leave a comment to remove it when we inplement trusty circle

**Comment by vijeth-aradhya** [comment]
Why is travis failing now :tired_face: I don't think it's because of any of my additions though..

**Comment by jayvdb** [comment]
unack 79f4711

File: bears/haskell/GhcModBear.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+                     r'(?P<message>.+)')
+class GhcModBear:
+    """
+    Syntax checking with ghc for haskell files.
+
+    See <https://hackage.haskell.org/package/ghc-mod> for more information!
+    """
+
+    LANGUAGES = {'Haskell'}
+    REQUIREMENTS = {DistributionRequirement(apt_get='ghc-mod')}
```
Comment (Mixih) [inline bears/haskell/GhcModBear.py]: I think you specify the cabal dep here.

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: yess!! i'll do that when we update `dependecy_management`! :smile:

File: bears/haskell/GhcModBear.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    DistributionRequirement)
+
+
+@linter(executable='ghc-mod',
+        output_format='regex',
+        output_regex=r'.+:(?P<line>\d+):(?P<column>\d+):'
+                     r'(?P<message>.+)')
+class GhcModBear:
+    """
+    Syntax checking with ghc for haskell files.
```
Comment (Makman2) [inline bears/haskell/GhcModBear.py]: I think you should write first letter of `Haskell` with upper-case, they do this too on their page^^

File: tests/haskell/GhcModBearTest.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+@generate_skip_decorator(GhcModBear)
+class GhcModBearTest(LocalBearTestHelper):
+
+    def setUp(self):
+        self.section = Section('name')
+        self.uut = GhcModBear(self.section, Queue())
+
+    def test_valid(self):
+        self.check_validity(self.uut, hello_world_file.split('\n'), valid=True,
```
Comment (Makman2) [inline tests/haskell/GhcModBearTest.py]: use `splitlines()` instead of `split('\n')` :)

Comment (Makman2) [inline tests/haskell/GhcModBearTest.py]: actually you could do that already for your strings instead of doing it here :+1:

File: .ci/deps.shellcheck.sh [hunk 15]
```diff
@@ -24,3 +24,6 @@ else
     echo "Using cached ShellCheck $EXISTING_VERSION"
   fi
 fi
+
+cabal update && cabal install cabal-install-1.22.9.0
+cabal install ghc-mod
```
Comment (Makman2) [inline .ci/deps.shellcheck.sh]: oh I think this should go to `deps.sh` instead of this spellcheck install script^^

Comment (Mixih) [inline .ci/deps.shellcheck.sh]: agreed, WIP because of this

File: tests/haskell/GhcModBearTest.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+from coalib.testing.BearTestHelper import generate_skip_decorator
+from coalib.settings.Section import Section
+from coalib.settings.Setting import Setting
+
+
+# A simple hello world program!
+hello_world_file = """
+main :: IO()
+main = putStrLn "Hello World!"
+"""
```
Comment (Mixih) [inline tests/haskell/GhcModBearTest.py]: for example you could change these to `""".splitLines()` to implement what makamn2 advised

File: bear-requirements.txt [hunk 17]
```diff
@@ -29,3 +29,4 @@ vim-vint~=0.3.10
 vulture~=0.10.0
 yamllint~=1.5
 yapf~=0.14.0
+dependency_management~=0.2.0
```
Comment (Mixih) [inline bear-requirements.txt]: Please don't add this, it can cause all kinds of dependency problems as we've already inherited it from coala. I see you have a pull there, Please remove this when it gets accepted.

**Comment by sils** [comment]
@vijeth-aradhya you'll have to update coala in requirements.txt to version `>=0.10.0.dev20170106003439` to pick up the change you did

File: bears/haskell/GhcModBear.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+    See <https://hackage.haskell.org/package/ghc-mod> for more information!
+    """
+
+    LANGUAGES = {'Haskell'}
+    REQUIREMENTS = {CabalRequirement(package='ghc-mod', version='5.6.0')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/97863'
```
Comment (sils) [inline bears/haskell/GhcModBear.py]: FWIW it looks like there's spaces missing in the message, is that like that in the output of the tool or a bug of ours somewhere?

**Comment by sils** [comment]
tests are much better, I feel like the extra round of review really added some quality here :+1: I hope this wasn't too frustrating :/. If you change the requirements as indicated above I think it'll work and we can *really* finally merge it.

File: bears/haskell/GhcModBear.py (hunk 18 above)
Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: spaces missing?! What do you mean :o

Comment (sils) [inline bears/haskell/GhcModBear.py]: `from the literal ‘3’In the first` no space before In

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: hmm i'll check it out now :+1:

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: This is `ghc-7.10.3` output on my local machine for the first testfile
```
/GhcModBearTests/testfile1.hs:2:17:
    No instance for (Num String) arising from the literal ‘3’
    In the first argument of ‘putStrLn’, namely ‘3’
    In the expression: putStrLn 3
    In an equation for ‘main’: main = putStrLn 3
```
Looks to me that wherever there is `\n`, it's just joining the strings

Comment (sils) [inline bears/haskell/GhcModBear.py]: can you file an issue for linter in coala?

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: It'll be nice if we make it prettier I guess :+1: Hmm anyway we can add a `\n` there?!

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Oh wait that is `ghc` output, this is `ghc-mod-5.6.0` output
There is nothing wrong in out Linter - it's their output which is like that :+1: 
```
~/GhcModBearTests/testfile1.hs:2:17:No instance for (Num String) arising from the literal ‘3’In the first argument of ‘putStrLn’, namely ‘3’In the expression: putStrLn 3In an equation for ‘main’: main = putStrLn 3
```

Comment (sils) [inline bears/haskell/GhcModBear.py]: cool, file a bug upstream ok?

Comment (vijeth-aradhya) [inline bears/haskell/GhcModBear.py]: Yup I did, [here](https://github.com/DanielG/ghc-mod/issues/863) :+1:

**Comment by vijeth-aradhya** [comment]
Travis still weirdly failing .. :disappointed:

**Comment by jayvdb** [comment]
Completed at https://github.com/coala/coala-bears/pull/1267
